src/dcast.d

```diff
@@ -75,6 +75,20 @@ extern (C++) Expression implicitCastTo(Expression e, Scope* sc, Type t)
                     result.type = t;
                     return;
                 }
+                if (match == MATCHconvert && sc.func && e.type.ty == Tarray && t.ty == Tarray)
+                {
+                    /* Issue 15702: @safe code should not allow implicit
+                     * conversion of T[] to void[] if T has indirections. */
+                    auto et = e.type.nextOf();
+                    auto tn = t.nextOf();
+                    if (et.hasPointers() && tn.ty == Tvoid && sc.func.setUnsafe())
+                    {
+                        e.error("cannot implicitly convert expression (%s) of type %s with indirections to %s in @safe code",
```

Comment:
Fixed the error message.

And you're right, I missed the `const(void)[]` case. Fixed.